pandas/core/groupby/grouper.py

```diff
@@ -413,7 +413,6 @@ def _set_grouper(self, obj: NDFrame, sort: bool = False):
         # "NDFrameT", variable has type "None")
         self.obj = obj  # type: ignore[assignment]
         self._gpr_index = ax
```

Comment:
if you can find a way to avoid setting these attributes instead, that'd make this class easier to reason about

Reply:
I agree, if I understand what you're saying. Do you mean make this whole class less stateful?

I actually started doing this because I thought the same thing. I wanted to make this Grouper be immutable. Make this class, Grouper, not know anything about `obj`, and then a separate class BoundGrouper, that is the pairing of a Grouper and a specific `obj`. Does that sound like a reasonable way to organize things?

Reply:
Yeah I was worried about user facing changes. What exactly do we consider user-facing? In the docs Grouper is pretty much a namedtuple, and I would argue this should be the public API (just the attributes given in the constructor).

What gets in the way of this are the `ax` and `groups` properties, which look like they are supposed to be public. But I'm pretty sure it is a mistake to expose those two, since they depend on the Grouper being bound to an obj, and there isn't a user-facing way to bind a Grouper to an obj. So I feel like any user who is relying on those two properties is already reaching too far into the Grouper and therefore that should be considered undefined.

Thoughts?